libr/io/p/io_gprobe.c

```diff
@@ -47,12 +58,125 @@ enum {
 	GPROBE_ACK = 0x0c,
 	GPROBE_RESET = 0x20,
 	GPROBE_GET_DEVICE_ID = 0x30,
+	GPROBE_GET_INFORMATION = 0x40,
 	GPROBE_RAM_READ_2 = 0x52,
 	GPROBE_RAM_WRITE_2 = 0x53,
 	GPROBE_RUN_CODE_2 = 0x54,
 };
 
-static int sp_close (struct sp_port *port) {
+#if __UNIX__
+static ut8 gprobe_checksum_i2c (const ut8 *p, unsigned int size, ut8 initial) {
+	ut8 res = initial;
+	unsigned int k;
+
+	for (k = 0; k < size; ++k)
+		res ^= p[k];
+
+	return res;
+}
+
+static void gprobe_frame_i2c (RBuffer *frame) {
+	ut8 size = r_buf_size (frame) + 1;
+	ut8 header[] = {0x51, 0x80 + size + 3, 0xc2, 0x00, 0x00};
+	ut8 checksum;
+
+	r_buf_prepend_bytes (frame, &size, 1);
+	r_buf_prepend_bytes (frame, header, sizeof (header));
+
+	checksum = gprobe_checksum_i2c (r_buf_buffer (frame), r_buf_size (frame), GPROBE_I2C_ADDR);
+
+	r_buf_append_bytes (frame, &checksum, 1);
+}
+
+static int gprobe_get_reply_i2c (struct gport *port, ut8 cmd, RBuffer *reply) {
+	ut8 buf[131];
+	int count;
+	int ddc2bi3_len;
+	ut8 addr = 0x50;
+	ut8 checksum;
+	struct timespec delay_40ms = {0, 40000000};
+
+	nanosleep (&delay_40ms, NULL);
+
+	count = read (port->fd, buf, sizeof (buf));
+
+	if (count != sizeof (buf))
+		return -1;
+
+	if ((buf[0] & 0xfe) != GPROBE_I2C_ADDR)
+		return -1;
+
+	if (!(buf[1] & 0x80))
+		return -1;
+
+	ddc2bi3_len = buf[1] & ~0x80;
+
+	if (buf[2] != 0xc2)
+		return -1;
+
+	if (buf[3] != 0x00)
+		return -1;
+
+	if (buf[4] != 0x00)
+		return -1;
+
+	if (cmd != buf[6])
+		return -1;
+
+	if (!(buf[5] - 2))
+		return 0;
+
+	if (buf[5] != ddc2bi3_len - 2)
+		return -1;
+
+	checksum = gprobe_checksum_i2c (&addr, 1, 0);
+
+	if (gprobe_checksum_i2c (buf, ddc2bi3_len + 2, checksum) != buf[ddc2bi3_len + 2])
+		eprintf ("gprobe rx checksum error\n");
+
+	r_buf_append_bytes (reply, buf + 7, buf[5] - 3);
+
+	return 0;
+}
+
+static int gprobe_send_request_i2c (struct gport *port, RBuffer *request) {
+	if (write (port->fd, r_buf_buffer (request), r_buf_size (request)) != r_buf_size (request))
+		return -1;
+
+	return 0;
+}
+
+static int i2c_open (struct gport *port) {
+	char filename[20];
+	char *end;
+	int i2cbus = strtol (port->name + 4, &end, 0);
+	int file;
+
+	if (*end)
+		return -1;
+
+	snprintf (filename, sizeof (filename), "/dev/i2c/%d", i2cbus);
+	filename[sizeof (filename) - 1] = '\0';
+	file = r_sandbox_open (filename, O_RDWR, 0);
+
+	if (file < 0 && (errno == ENOENT || errno == ENOTDIR)) {
+		sprintf (filename, "/dev/i2c-%d", i2cbus);
+		file = r_sandbox_open (filename, O_RDWR, 0);
+	}
+
+	if (file < 0)
+		return -1;
+
+	if (ioctl (file, I2C_SLAVE, GPROBE_I2C_ADDR >> 1) < 0)
```

Comment:
braces, and ioctl() is unix only. that wont work on windows. so maybe we should just #if __UNIX__ the whole plugin, but it will be good to know if this thing can run on windows somehow

Reply:
The i2c part is #if __UNIX__.
The serial part runs nicely under windows.